### PR TITLE
Add ExceptionState to ProgramState

### PR DESCRIPTION
**Pull request by pavel-mikula-sonarsource**
Related to #5710, prerequisite for #5718

We need to track what kind of exception are we dealing with.

We will need this to know that we're on an exception-thrown path for nested scenario.
```C#
try{
  try{
     throw new Exception(); // Persist state, that we're in an exception
  } finally {
    // After this, we need to skip DoSomething()
  }
  DoSomething();
} finally {
    // And continue with this
}
```

### CODE REVIEW

File: analyzers/src/SonarAnalyzer.Common/SymbolicExecution/Roslyn/ExceptionState.cs [hunk 1]
```diff
@@ -0,0 +1,40 @@
...
+ * along with this program; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+ */
+
+using System;
+using Microsoft.CodeAnalysis;
+
+namespace SonarAnalyzer.SymbolicExecution.Roslyn
+{
+    public sealed record ExceptionState : IEquatable<ExceptionState>
```
Comment (costin-zaharia-sonarsource) [inline analyzers/src/SonarAnalyzer.Common/SymbolicExecution/Roslyn/ExceptionState.cs:26]: I think we don't need to implement `IEquatable<ExceptionState>` here. It is part of record already.

File: analyzers/tests/SonarAnalyzer.UnitTest/SymbolicExecution/Roslyn/ExceptionStateTest.cs [hunk 2]
```diff
@@ -0,0 +1,74 @@
...
+ * along with this program; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+ */
+
+using SonarAnalyzer.SymbolicExecution.Roslyn;
+
+namespace SonarAnalyzer.UnitTest.SymbolicExecution.Roslyn
+{
+    [TestClass]
+    public partial class ExceptionStateTest
```
Comment (costin-zaharia-sonarsource) [inline analyzers/tests/SonarAnalyzer.UnitTest/SymbolicExecution/Roslyn/ExceptionStateTest.cs:26]: `partial` keyword is not needed.

File: analyzers/tests/SonarAnalyzer.UnitTest/SymbolicExecution/Roslyn/ExceptionStateTest.cs [hunk 3]
```diff
@@ -0,0 +1,74 @@
...
+        [TestMethod]
+        public void ToString_Unknown() =>
+            ExceptionState.UnknownException.ToString().Should().Be("Unknown");
+
+        [DataTestMethod]
+        [DataRow("System.Exception", "Exception")]
+        [DataRow("System.ArgumentNullException", "ArgumentNullException")]
+        [DataRow("System.IO.IOException", "IOException")]
+        public void ToString_Known(string typeName, string expected) =>
+            new ExceptionState(TestHelper.CompileCS(string.Empty).Model.Compilation.GetTypeByMetadataName(typeName)).ToString().Should().Be(expected);
```
Comment (costin-zaharia-sonarsource) [inline analyzers/tests/SonarAnalyzer.UnitTest/SymbolicExecution/Roslyn/ExceptionStateTest.cs:41]: To string will be used only for debugging? We could have wrong results when comparing exceptions with the same name from different namespaces.

**Review by costin-zaharia-sonarsource — CHANGES_REQUESTED** [review]

File: analyzers/tests/SonarAnalyzer.UnitTest/SymbolicExecution/Roslyn/ExceptionStateTest.cs (hunk 3 above)
Comment (pavel-mikula-sonarsource) [reply analyzers/tests/SonarAnalyzer.UnitTest/SymbolicExecution/Roslyn/ExceptionStateTest.cs:41]: It should be used only for debugging and debugging serialization (in ProgramState).

Branching decisions must be made on the `Type` property.

Comment (costin-zaharia-sonarsource) [reply analyzers/tests/SonarAnalyzer.UnitTest/SymbolicExecution/Roslyn/ExceptionStateTest.cs:41]: Discussed offline and it's preferred to have it as a string.

**Review by costin-zaharia-sonarsource — APPROVED** [review]
LGTM!

**Pull request merged**